src/store/Controller.cc

```diff
@@ -357,11 +357,12 @@ Store::Controller::allowSharing(StoreEntry &entry, const cache_key *key)
             throw TexcHere("cannot sync");
         if (!found) {
             if (!transients->hasWriter(entry)) {
-                // Cannot anchor right now and will not be able in the future.
-                // Prevent others from falling into the same trap.
+                // prevent others from falling into the same trap
                 stopSharing(entry);
-                throw TexcHere("unattached transients entry");
-            } else if (!entry.hittingRequiresCollapsing()) {
+                throw TexcHere("unattached transients entry missing writer");
+            }
+
+            if (!entry.hittingRequiresCollapsing()) {
```

Comment:
Should we move this BUG check up, before we may bail due to the just-disappeared writer? It feels like the lack of a hittingRequiresCollapsing() flag for a `!found` entry is a bug, regardless of whether the writer has disappeared after the Transients::get() call or not. Am I missing something here? Perhaps this move requires some other pending changes related to the hittingRequiresCollapsing() flag?

Reply:
> whether the writer has disappeared after the Transients::get() call or not.

Also, there could be no writer at Transients::get() time, I think we cannot distinguish these two situations. If there was no writer, hittingRequiresCollapsing() remains false and there is no bug here. That is why I check for a writer first.  After that,  we can assert that the existing writer implies hittingRequiresCollapsing().

Reply:
I can rephrase my concern as the following expectation: `!found` should imply hittingRequiresCollapsing(). This expectation is the strongest upon successfully leaving allowSharing(), but it would probably be best to preserve that invariant at false anchorToCache() exit time as well. My expectation may be wrong, but, if it is wrong, then, unless I am missing something obvious, the code should not suggest such a wrong expectation in the first place.

Reply:
>  there could be no writer at Transients::get() time

I think this hypothetical situation could happen in a case of a bug in the newly added StoreMap::closeForReadingAndFreeIdle() and other routines responsible for that cleanup. Anyway, even if such a situation happens, checking for hittingRequiresCollapsing() first should reveal that problem (as well as the previous hasWriter() check), so I do not mind moving the 'BUG' check up (done at c531911). I had to split the !isWriter condition into two: I think we should always stopSharing() in this case, even if hittingRequiresCollapsing() check fails too.

Reply:
I see two red flags in the current implementation, and addressing the second concern will create the third red flag:

1. The current (stopSharing(), check hittingRequiresCollapsing()) order implies that an unshareable entry may still be collapsed on. While the current implementation of those two methods may allow for any call order, that is likely a side effect of a hidden problem with one or both methods rather than a fundamental invariant that is likely to survive future code adjustments. To avoid this implication, we should check hittingRequiresCollapsing() (or remember its value) _before_ calling stopSharing().
2. Calling stopSharing() from startSharing() implies that we want to undo (some of the) startSharing() effects. I think this is a good idea, but it looks strange that we only stopSharing() upon discovery of some, but not necessarily all sharing problems. To address this concern, consider moving the stopSharing() call into find(), where allowSharing() is called, and where exceptions are already handled.
3. When handling exceptions, find() keeps the entry shareable (i.e. it calls release() with a true flag). In this PR, we want to call stopSharing() when throwing an exception. Please double check whether that release() parameter in find() should actually be false. And, if it should be false, please check whether we should call stopSharing() from StoreEntry::release(false) instead of doing that directly from the find()'s `catch` clause.

Reply:
I tried to address your concerns at c859658.

> we should check hittingRequiresCollapsing() (or remember its value) before calling stopSharing()

I agree: since after stopSharing() has been called no more 'hitting' is possible on this entry, 'collapsing' simply does not have any sense.

> consider moving the stopSharing() call into find(), where allowSharing() is called

This concern looks reasonable: for symmetry/simplicity sake it is better to call stopSharing()/allowSharing() at one level, than one from another.

> Please double check whether that release() parameter in find() should actually be false

That release() parameter effectively sets the StoreEntry::shareableWhenPrivate flag value. I think it should be false, at least for the two cases in question:
  * unexpected !hittingRequiresCollapsing() indicate an invalid (and unrecoverable) entry state and BUG report does warn about that. We should stop sharing this entry even for already attached readers.
  * unexpected !hasWriter means that no future data notifications will happen and, hence, no sharing for already attached readers is possible (for otherwise, they will hang forever).

We can probably formulate a general rule that any exception in Controller::allowSharing() scope indicates that the entry is not suitable for any sharing, including 'private sharing'.


> whether we should call stopSharing() from StoreEntry::release(false)

We should not call stopSharing() from there because it is already performed (implicitly): stopSharing() effectively means Transients::evictCached(), whereas release() calls Controller::evictCached(), which evicts the entry from all caches.

Reply:
> We should not call stopSharing() from release() because it is already performed (implicitly): [both call] evictCached()

Yes, release() appears to call evictCached(), even on the releaseRequest() path, because our StoreEntry should have a key -- we call addReading() earlier in allowSharing().

I think this PR should be content with the `release(false)` call. Compared to release(false), stopSharing() looks like a "milder" sharing-focused action that may not do enough to fully get rid of the problematic entry. There may be more, deeper problems here, but no obvious ones that fall into this PR scope. I am resolving this change request.